File: results/dnd-rest-benchmark/lifecycle-runs/20260731T060905Z_claude_claude-sonnet-5_python-django/dndsite/views/compendium.py
"""Monster and item reference data: create + fetch by slug."""

import json
import re

from django.http import JsonResponse
from django.views.decorators.csrf import csrf_exempt

from dndsite import db
from dndsite.rules import numeric
from dndsite.views._util import json_body

SLUG_RE = re.compile(r"^[a-z0-9-]+$")


def valid_slug(slug):
    return isinstance(slug, str) and bool(SLUG_RE.match(slug))


@csrf_exempt
def monsters_collection(request):
    if request.method != "POST":
        return JsonResponse({"error": "method not allowed"}, status=405)
    try:
        body = json_body(request)
        slug = body["slug"]
        name = body["name"]
        cr = body["cr"]
        armor_class = body["armor_class"]
        hit_points = body["hit_points"]
        tags = body.get("tags", [])
    except (json.JSONDecodeError, KeyError, TypeError):
        return JsonResponse({"error": "invalid request"}, status=400)

    if not valid_slug(slug):
        return JsonResponse({"error": "invalid slug"}, status=400)
    if not isinstance(name, str) or not name:
        return JsonResponse({"error": "invalid name"}, status=400)
    if not isinstance(cr, str) or not cr:
        return JsonResponse({"error": "invalid cr"}, status=400)
    if not isinstance(armor_class, int) or isinstance(armor_class, bool):
        return JsonResponse({"error": "invalid armor_class"}, status=400)
    if not isinstance(hit_points, int) or isinstance(hit_points, bool):
        return JsonResponse({"error": "invalid hit_points"}, status=400)
    if not isinstance(tags, list) or not all(isinstance(t, str) for t in tags):
        return JsonResponse({"error": "invalid tags"}, status=400)

    if db.get_monster(slug) is not None:
        return JsonResponse({"error": "monster already exists"}, status=409)

    monster = {
        "slug": slug,
        "name": name,
        "cr": cr,
        "armor_class": armor_class,
        "hit_points": hit_points,
        "tags": tags,
    }
    db.create_monster(monster)

    return JsonResponse(
        {
            "slug": slug,
            "name": name,
            "cr": cr,
            "armor_class": armor_class,
            "hit_points": hit_points,
        },
        status=201,
    )


def monster_detail(request, slug):
    if request.method != "GET":
        return JsonResponse({"error": "method not allowed"}, status=405)

    monster = db.get_monster(slug)
    if monster is None:
        return JsonResponse({"error": "monster not found"}, status=404)

    return JsonResponse(monster)


@csrf_exempt
def items_collection(request):
    if request.method != "POST":
        return JsonResponse({"error": "method not allowed"}, status=405)
    try:
        body = json_body(request)
        slug = body["slug"]
        name = body["name"]
        item_type = body["type"]
        rarity = body["rarity"]
        cost_gp = body["cost_gp"]
    except (json.JSONDecodeError, KeyError, TypeError):
        return JsonResponse({"error": "invalid request"}, status=400)

    if not valid_slug(slug):
        return JsonResponse({"error": "invalid slug"}, status=400)
    if not isinstance(name, str) or not name:
        return JsonResponse({"error": "invalid name"}, status=400)
    if not isinstance(item_type, str) or not item_type:
        return JsonResponse({"error": "invalid type"}, status=400)
    if not isinstance(rarity, str) or not rarity:
        return JsonResponse({"error": "invalid rarity"}, status=400)
    if not isinstance(cost_gp, (int, float)) or isinstance(cost_gp, bool):
        return JsonResponse({"error": "invalid cost_gp"}, status=400)

    if db.get_item(slug) is not None:
        return JsonResponse({"error": "item already exists"}, status=409)

    item = {
        "slug": slug,
        "name": name,
        "type": item_type,
        "rarity": rarity,
        "cost_gp": numeric(cost_gp),
    }
    db.create_item(item)

    return JsonResponse(item, status=201)


def item_detail(request, slug):
    if request.method != "GET":
        return JsonResponse({"error": "method not allowed"}, status=405)

    item = db.get_item(slug)
    if item is None:
        return JsonResponse({"error": "item not found"}, status=404)

    return JsonResponse(item)

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T062307Z_claude_claude-sonnet-5_typescript-node/src/routes/compendium.ts
// Monster and item reference data (create + lookup by slug). `getMonster`
// is also used by routes/dm.ts to resolve monster slugs for encounter
// building.
import type { ServerResponse } from "node:http";
import { db } from "../db.js";
import { sendJson } from "../http.js";
import { isPlainObject, isValidInt, SLUG_RE } from "../validation.js";

export interface MonsterRecord {
  slug: string;
  name: string;
  cr: string;
  armorClass: number;
  hitPoints: number;
  tags: string[];
}

function hasMonster(slug: string): boolean {
  const row = db.prepare("SELECT 1 FROM monsters WHERE slug = ?").get(slug);
  return row !== undefined;
}

export function getMonster(slug: string): MonsterRecord | undefined {
  const row = db
    .prepare("SELECT slug, name, cr, armor_class, hit_points, tags FROM monsters WHERE slug = ?")
    .get(slug) as
    | { slug: string; name: string; cr: string; armor_class: number; hit_points: number; tags: string }
    | undefined;
  if (!row) return undefined;
  return {
    slug: row.slug,
    name: row.name,
    cr: row.cr,
    armorClass: row.armor_class,
    hitPoints: row.hit_points,
    tags: JSON.parse(row.tags) as string[],
  };
}

function saveMonster(monster: MonsterRecord): void {
  db.prepare(
    "INSERT INTO monsters (slug, name, cr, armor_class, hit_points, tags) VALUES (?, ?, ?, ?, ?, ?)",
  ).run(monster.slug, monster.name, monster.cr, monster.armorClass, monster.hitPoints, JSON.stringify(monster.tags));
}

export function handleCreateMonster(res: ServerResponse, body: unknown): void {
  if (
    !isPlainObject(body) ||
    typeof body.slug !== "string" ||
    typeof body.name !== "string" ||
    typeof body.cr !== "string" ||
    !isValidInt(body.armor_class, 0, Number.MAX_SAFE_INTEGER) ||
    !isValidInt(body.hit_points, 0, Number.MAX_SAFE_INTEGER) ||
    !Array.isArray(body.tags) ||
    !body.tags.every((tag) => typeof tag === "string")
  ) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  if (!SLUG_RE.test(body.slug)) {
    sendJson(res, 400, { error: "invalid slug" });
    return;
  }

  if (hasMonster(body.slug)) {
    sendJson(res, 409, { error: "monster already exists" });
    return;
  }

  const monster: MonsterRecord = {
    slug: body.slug,
    name: body.name,
    cr: body.cr,
    armorClass: body.armor_class,
    hitPoints: body.hit_points,
    tags: body.tags as string[],
  };
  saveMonster(monster);

  sendJson(res, 201, {
    slug: monster.slug,
    name: monster.name,
    cr: monster.cr,
    armor_class: monster.armorClass,
    hit_points: monster.hitPoints,
  });
}

export function handleGetMonster(res: ServerResponse, slug: string): void {
  const monster = getMonster(slug);
  if (!monster) {
    sendJson(res, 404, { error: "monster not found" });
    return;
  }

  sendJson(res, 200, {
    slug: monster.slug,
    name: monster.name,
    cr: monster.cr,
    armor_class: monster.armorClass,
    hit_points: monster.hitPoints,
    tags: monster.tags,
  });
}

interface ItemRecord {
  slug: string;
  name: string;
  type: string;
  rarity: string;
  costGp: number;
}

function hasItem(slug: string): boolean {
  const row = db.prepare("SELECT 1 FROM items WHERE slug = ?").get(slug);
  return row !== undefined;
}

function getItem(slug: string): ItemRecord | undefined {
  const row = db.prepare("SELECT slug, name, type, rarity, cost_gp FROM items WHERE slug = ?").get(slug) as
    | { slug: string; name: string; type: string; rarity: string; cost_gp: number }
    | undefined;
  if (!row) return undefined;
  return { slug: row.slug, name: row.name, type: row.type, rarity: row.rarity, costGp: row.cost_gp };
}

function saveItem(item: ItemRecord): void {
  db.prepare("INSERT INTO items (slug, name, type, rarity, cost_gp) VALUES (?, ?, ?, ?, ?)").run(
    item.slug,
    item.name,
    item.type,
    item.rarity,
    item.costGp,
  );
}

export function handleCreateItem(res: ServerResponse, body: unknown): void {
  if (
    !isPlainObject(body) ||
    typeof body.slug !== "string" ||
    typeof body.name !== "string" ||
    typeof body.type !== "string" ||
    typeof body.rarity !== "string" ||
    !isValidInt(body.cost_gp, 0, Number.MAX_SAFE_INTEGER)
  ) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  if (!SLUG_RE.test(body.slug)) {
    sendJson(res, 400, { error: "invalid slug" });
    return;
  }

  if (hasItem(body.slug)) {
    sendJson(res, 409, { error: "item already exists" });
    return;
  }

  const item: ItemRecord = {
    slug: body.slug,
    name: body.name,
    type: body.type,
    rarity: body.rarity,
    costGp: body.cost_gp,
  };
  saveItem(item);

  sendJson(res, 201, {
    slug: item.slug,
    name: item.name,
    type: item.type,
    rarity: item.rarity,
    cost_gp: item.costGp,
  });
}

export function handleGetItem(res: ServerResponse, slug: string): void {
  const item = getItem(slug);
  if (!item) {
    sendJson(res, 404, { error: "item not found" });
    return;
  }

  sendJson(res, 200, {
    slug: item.slug,
    name: item.name,
    type: item.type,
    rarity: item.rarity,
    cost_gp: item.costGp,
  });
}
